pallets/tinlake-investor-pool/src/lib.rs

```diff
@@ -740,6 +863,106 @@ pub mod pallet {
 			};
 		}
 
+		pub(crate) fn calculate_collect(
+			loc: TrancheLocator<T::PoolId, T::TrancheId>,
+			order: UserOrder<T::Balance, T::EpochId>,
+			pool: PoolDetails<T::AccountId, T::CurrencyId, T::EpochId, T::Balance>,
+			end_epoch: T::EpochId,
+		) -> OutstandingCollections<T::Balance> {
+			let mut epoch_idx = order.epoch;
+
+			// No collect possible in this epoch
+			if epoch_idx == pool.current_epoch {
+				return OutstandingCollections {
+					payout_currency_amount: Zero::zero(),
+					payout_token_amount: Zero::zero(),
+					remaining_supply_currency: order.supply,
+					remaining_redeem_token: order.redeem,
+				};
+			}
+
+			// It is only possible to collect epochs which are already over
+			let parse_until_epoch = if end_epoch > pool.last_epoch_executed {
+				pool.last_epoch_executed
+			} else {
+				end_epoch
+			};
+
+			let mut payout_currency_amount: T::Balance = Zero::zero();
+			let mut payout_token_amount: T::Balance = Zero::zero();
+			let mut remaining_supply_currency = order.supply;
+			let mut remaining_redeem_token = order.redeem;
+			while epoch_idx <= parse_until_epoch
+				&& (remaining_supply_currency != Zero::zero()
+					|| remaining_redeem_token != Zero::zero())
+			{
+				if remaining_supply_currency != Zero::zero() {
```

Comment:
Are the two `if`-clauses related in anyway? If not we could split them into two separate functions. Makes the `&& (remaining_supply_currency != Zero::zero() || remaining_redeem_token != Zero::zero())` go away and overall more readable.

Reply:
I would parse the `epoch` as a parameter, as each `if`-clause currently accesses storage separately which is costly.